src/QirRuntime/lib/Tracer/tracer.cpp

```diff
@@ -0,0 +1,305 @@
+    void CTracer::AddOperationToLayer(OpId id, LayerId layer)
+    {
+        assert(layer < this->metricsByLayer.size());
+        this->metricsByLayer[layer].operations[id] += 1;
+    }
+
+    //------------------------------------------------------------------------------------------------------------------
+    // CTracer::UpdateQubitState
+    //------------------------------------------------------------------------------------------------------------------
+    void CTracer::UpdateQubitState(Qubit q, LayerId layer, Duration opDuration)
+    {
+        QubitState& qstate = this->UseQubit(q);
+        for (OpId idPending : qstate.pendingZeroOps)
+        {
+            this->AddOperationToLayer(idPending, layer);
+        }
+
+        // Update the qubit state.
+        qstate.layer = layer;
+        const Time layerStart = this->metricsByLayer[layer].startTime;
+        qstate.lastUsedTime = max(layerStart, qstate.lastUsedTime) + opDuration;
+        qstate.pendingZeroOps.clear();
+    }
+
+    //------------------------------------------------------------------------------------------------------------------
+    // CTracer::TraceSingleQubitOp
+    //------------------------------------------------------------------------------------------------------------------
+    LayerId CTracer::TraceSingleQubitOp(OpId id, Duration opDuration, Qubit target)
+    {
+        this->seenOps.insert(id);
+
+        QubitState& qstate = this->UseQubit(target);
+        if (opDuration == 0 &&
+            (qstate.layer == INVALID || (this->globalBarrier != INVALID && qstate.layer < this->globalBarrier)))
+        {
+            qstate.pendingZeroOps.push_back(id);
+            return INVALID;
+        }
+
+        // Figure out the layer this operation should go into.
+        LayerId layerToInsertInto = this->FindLayerToInsertOperationInto(target, opDuration);
+        if (layerToInsertInto == INVALID)
+        {
+            layerToInsertInto = this->CreateNewLayer(opDuration);
+        }
+
+        // Add the operation and the pending zero-duration ones into the layer.
+        this->AddOperationToLayer(id, layerToInsertInto);
+        this->UpdateQubitState(target, layerToInsertInto, opDuration);
+
+        return layerToInsertInto;
+    }
+
+    //------------------------------------------------------------------------------------------------------------------
+    // CTracer::TraceControlledSingleQubitOp
+    //------------------------------------------------------------------------------------------------------------------
+    LayerId CTracer::TraceMultiQubitOp(
+        OpId id,
+        Duration opDuration,
+        long nFirstGroup,
+        Qubit* firstGroup,
+        long nSecondGroup,
+        Qubit* secondGroup)
+    {
+        assert(nFirstGroup >= 0);
+        assert(nSecondGroup > 0);
+
+        this->seenOps.insert(id);
+
+        // Operations that involve a single qubit can special case duration zero.

```

Comment:
Something is skipped in the comment...
> // Operations that involve a single qubit **can special case** duration zero.

Reply:
No, it's me taking liberty with English and using "special case" as a verb. Moved the comments around to make it clear (hopefully) _why_ we bother to redirect to single qubit case.